### ISSUE TITLE
Dropout is broken

### ISSUE THREAD

**Issue by Refefer** [lisa-lab/pylearn2#1387]
Primer: https://groups.google.com/forum/#!topic/pylearn-users/WqkLMsf_sjU

I've noticed that dropout causes major deviations from the objective function, given enough time, for different network topologies.  More worrisome is that rather than stagnating the validation set misclassification error, it will reverse course on both the train set and validation set.

I have tested pylearn2 onmaster, a year old pylearn2 and found both exhibit the same problem.  Testing on gpu or cpu with float32/float64 doesn't fix the problem.  What's worse, while I originally stated that having sparse initializations seemed to mitigate the problem, given enough time I've found the objective deviates from the cost function.

I should also mention I've tested it on Theano 0.6.0 as well as master.

Any ideas what's going on?

**Comment by Refefer** [comment]
I should also mention I've tested it on Theano 0.6.0 as well as master.

**Comment by dwf** [comment]
So `_objective` and `_nll` are supposed to be different.

`_objective` is added by the SGD algorithm and it's a measurement of the actual `Cost` being minimized. In this case, it's a stochastic one, with a different dropout mask applied each time. In the case of dropout it's mildly useless to actually compute this quantity, but SGD is a generic thing, it has no idea that the thing it's minimizing is stochastic, so it just adds a channel for it.

`_nll` is added by the `Cost` and is the log likelihood of the predictions of the model without dropout/scaling applied.

That these two diverge is not terribly surprising.

**Comment by goodfeli** [comment]
In other words, _objective gives the average performance of a single sub-network, and _nll gives the performance of the ensemble of all sub-networks approximated using the weight scaling trick. The divergence between them shows the amount of benefit you get from using an ensemble method.
